cirq-rigetti/cirq_rigetti/quil_input_test.py

```diff
@@ -160,36 +181,149 @@ def test_quil_with_defgate():
+        sympy.Symbol("theta_1"),
+        sympy.Symbol("theta_2"),
+        sympy.Symbol("theta_3"),
+    )
+    cirq_circuit = Circuit(
+        [
+            XPowGate(exponent=1, global_shift=-0.5)(q0),
+            XPowGate(exponent=theta_0 / np.pi, global_shift=-0.5)(q1),
+            XPowGate(exponent=(2 / np.pi) * theta_1, global_shift=-0.5)(q3),
+            XPowGate(exponent=(2 / np.pi) * theta_2 + 1 / np.pi, global_shift=-0.5)(q2),
+            XPowGate(
+                exponent=(2 / np.pi) * sympy.cos(theta_3) * sympy.exp(1j * theta_3),
+                global_shift=-0.5,
+            )(q4),
+        ]
+    )
+
+    assert cirq_circuit == circuit
+
+
 def test_measurement_without_classical_reg():
     """Measure operations must declare a classical register."""
     with pytest.raises(UnsupportedQuilInstruction):
         circuit_from_quil("MEASURE 0")
+
+
+# Insert a similar test for Kraus ops
+
+
+QUIL_PROGRAM_WITH_READOUT_NOISE = """
+DECLARE ro BIT[1]
+RX(pi) 0
+PRAGMA READOUT-POVM 0 "(0.9 0.050000000000000044 0.09999999999999998 0.95)"
+MEASURE 0 ro[0]
+"""
+
+
+def test_readout_noise():
+    """Convert a program with readout noise."""
+    program = Program(QUIL_PROGRAM_WITH_READOUT_NOISE)
+    circuit = circuit_from_quil(program)
+
+    result = Simulator(seed=0).run(circuit, repetitions=1000)
+    assert result.histogram(key="ro[0]")[1] < 1000
+    assert result.histogram(key="ro[0]")[1] > 900
+
+
+def test_resolve_parameters():
+    """Test that parameters are correctly resolved for defined parametric gate."""
+    theta, beta = Parameter("theta"), Parameter("beta")
+    xy_matrix = np.array(
+        [
+            [1, 0, 0, 0],
+            [0, quil_cos(theta / 2), 1j * quil_sin(theta / 2) * quil_exp(1j * beta), 0],
+            [0, 1j * quil_sin(theta / 2) * quil_exp(1j * beta), 1j * quil_cos(theta / 2), 0],
+            [0, 0, 0, 1],
+        ]
+    )
+
+    defgate = DefGate("PHASEDXY", xy_matrix, parameters=[beta, theta])
+
+    cirq_phased_xy = defgate_to_cirq(defgate)
+
+    op = cirq_phased_xy(beta=sympy.Symbol("beta"), theta=sympy.Symbol("theta"))(
+        cirq.LineQubit(0), cirq.LineQubit(1)
+    )
+
+    op._resolve_parameters_({"beta": 1.0, "theta": 2.0}, True)
+
+
+def test_op_identifier():
+    """Check that converted parametric defgates will be correctly identified."""
+    theta, beta = Parameter("theta"), Parameter("beta")
+    xy_matrix = np.array(
+        [
+            [1, 0, 0, 0],
+            [0, quil_cos(theta / 2), 1j * quil_sin(theta / 2) * quil_exp(1j * beta), 0],
+            [0, 1j * quil_sin(theta / 2) * quil_exp(1j * beta), 1j * quil_cos(theta / 2), 0],
+            [0, 0, 0, 1],
+        ]
+    )
+
+    defgate = DefGate("PHASEDXY", xy_matrix, parameters=[beta, theta])
+
+    gate1 = defgate_to_cirq(defgate)
+    gate2 = defgate_to_cirq(defgate)
+
+    op = gate1(beta=np.pi, theta=np.pi)(cirq.LineQubit(0), cirq.LineQubit(1))
+    op_id = cirq.OpIdentifier(gate2)
+
+    assert op in op_id
```

Comment:
This test is failing with this message: 
```
<abc.PHASEDXY object at 0x144e5e210>.on(cirq.LineQubit(0), cirq.LineQubit(1))

in

cirq.devices.noise_utils.OpIdentifier(abc.PHASEDXY, )
```

Inspecting the variables here, I'd expect this test to pass based on the implementation at https://github.com/quantumlib/Cirq/blob/main/cirq-core/cirq/ops/gateset.py#L226:
```py
op.gate
# <abc.PHASEDXY object at 0x13f8e9dd0>

op_id._gate_family.gate
# <class 'abc.PHASEDXY'>

Reply:
I don't understand how this could happen either.

Reply:
@bramathon  If I change the test to the following, only the second one fails

```py
    gate1 = defgate_to_cirq(defgate)
    gate2 = defgate_to_cirq(defgate)

    op = gate1(beta=np.pi, theta=np.pi)(cirq.LineQubit(0), cirq.LineQubit(1))

    assert op in cirq.OpIdentifier(gate1) # passes
    assert op in cirq.OpIdentifier(gate2) # fails
```

Looks like this is because [`defgate_to_cirq`](https://github.com/jselig-rigetti/Cirq/blob/rigetti-use-pyquil-v4/cirq-rigetti/cirq_rigetti/quil_input.py#L604-L613) is creating a whole new class each time it's called - I've added a `@cached_method` decorator on to `defgate_to_cirq` and it seems the tests are passing now. Let me know if this solution is undesirable